### PR TITLE
[FEAT] Adds a transform to perform the hashing trick

### LINKED ISSUE #164: Support the hashing trick as an encoding strategy for categorical features
For applications in sequential learning, it would be nice to have a stateful transform that doesn't presume every level of the categoricals is seen in the first chunk of training data. One approach used in sklearn is to include a FeatureHasher transformer, which I think would make sense here, too. 

The hashing trick would fit in cleanly with the existing C() API - potentially taking arguments for a hash seed and total_features or something. I'd be happy to contribute a PR for this if you think it'd make sense to include in formulaic.

### PR DESCRIPTION
**Pull request by rishi-kulkarni**
I went with the approach of not making `hashed` a stateful transform and just passing an empty dict to `encode_contrasts`. Is there any issue with that approach?

What sort of tests would you want to see here? Given the very pluggable nature of `formulaic` here, this feature ended up being very little code. 

Closes #164

### CODE REVIEW

File: formulaic/transforms/featurehash.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
+from hashlib import md5
```
Comment (matthewwardrop) [inline formulaic/transforms/featurehash.py:1]: nit: let's just call this module `hashed` to align with the transform name.

File: formulaic/transforms/featurehash.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    levels: int,
+    contrasts: Optional[
+        Union[Contrasts, Dict[str, Iterable[Number]], np.ndarray]
+    ] = None,
+    *,
+    hash_func: Callable[[str], int] = md5_to_int,
+    spans_intercept: bool = False,
+):
+    """
+    Hashes the values of a factor into a fixed number of levels. DUe to the
```
Comment (matthewwardrop) [inline formulaic/transforms/featurehash.py:27]: nit: typo: "DUe" -> "Due"

Comment (matthewwardrop) [inline formulaic/transforms/featurehash.py:27]: nit: Might be worth adding the word "Deterministically"? It is this deterministic nature that allows you to treat this as a regular function rather than a stateful transform.

File: formulaic/transforms/featurehash.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    large number of levels required to avoid collisions, this transform should
+    be used with output='sparse' to avoid memory issues.
```
Comment (matthewwardrop) [inline formulaic/transforms/featurehash.py:29]: note: If you *did* make it a stateful transform you could peek at the model spec and validate this explicitly, posting a warning if `output` is not `'sparse'`.

Comment (matthewwardrop) [inline formulaic/transforms/featurehash.py:29]: nit: code should be quoted using backticks.

**Review by matthewwardrop — CHANGES_REQUESTED** [review]
Hi @rishi-kulkarni . Thanks for the pull request. This looks pretty good. I left a few small comments.

Not in the comments, but probably the biggest gap, is that this assumes that the incoming data is a sequence of strings. Can you  explicitly do a type conversion rather than bailing out with some cryptic errors? It's fine if this is just in the hash function, since `str(...)` on a string should be pretty inexpensive.

**Comment by matthewwardrop** [comment]
> What sort of tests would you want to see here? Given the very pluggable nature of `formulaic` here, this feature ended up being very little code.

Aye... I'd be content with 100% coverage here. There are pretty few moving parts, and so long as every line is tested, and we also add a test for series of non-string types, I'm happy with this as is.

**Comment by rishi-kulkarni** [comment]
@matthewwardrop Thanks for the comments, they should now be addressed. I went with casting to `numpy.str_` since that seemed to have some performance benefits for very large numbers of rows.

File: formulaic/transforms/hashed.py [hunk 4]
```diff
@@ -24,9 +24,9 @@ def hashed(
     spans_intercept: bool = False,
 ):
     """
-    Hashes the values of a factor into a fixed number of levels. DUe to the
-    large number of levels required to avoid collisions, this transform should
-    be used with output='sparse' to avoid memory issues.
+    Deterministically the values of a factor into a fixed number of levels.
```
Comment (matthewwardrop) [inline formulaic/transforms/hashed.py:27]: nit: whoops. You dropped "hashes". Should be: "Deterministically hashes..."

**Review by matthewwardrop — APPROVED** [review]
Modulo some tiny docstring issues, this looks good. I'll merge once these changes have been made :).

File: formulaic/transforms/hashed.py (hunk 4 above)
Comment (rishi-kulkarni) [reply formulaic/transforms/hashed.py:27]: Good catch, so I did. Fixed.

**Comment by matthewwardrop** [comment]
I made one tiny further change to the docstring that I left in the review, but somehow it didn't propagate through. I'll merge this in now. Thank you so much for your contribution!

**Comment by rishi-kulkarni** [comment]
Hmm, looks like the `usedforsecurity` flag wasn't added until 3.9. I assume something like:

```
def md5_to_int(s: str) -> int:
    # Check the Python version
    if sys.version_info >= (3, 9):
        return int(md5(s.encode(), usedforsecurity=False).hexdigest(), 16)
    else:
        return int(md5(s.encode()).hexdigest(), 16)
```
should work
